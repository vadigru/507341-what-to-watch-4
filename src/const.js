export const MOVIES_DEFAULT_AMOUNT = 8;
export const MOVIES_STEP_AMOUNT = 8;
export const MAX_GENRES_AMOUNT = 9;
export const ALL_GENRES = `All genres`;
export const START_DELAY = 1000;

export const Tab = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const DefaultTab = {
  GENRE: `All genres`,
  MOVIE: `Overview`
};

export const RatingLevel = {
  ZERO: 0,
  BAD: 3,
  NORMAL: 5,
  GOOD: 8,
  AWESOME: 10
};

export const AppRoute = {
  ROOT: `/`,
  MOVIE_PAGE: `/movie`,
  SIGN_IN: `/login`,
  ADD_REVIEW: `/review`,
};
