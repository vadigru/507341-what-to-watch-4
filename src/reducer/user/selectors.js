import Namespace from "../namespace.js";

export const getAuthorizationStatus = (state) => {
  return state[Namespace.USER].authorizationStatus;
};

export const getInvalidAuthorizationStatus = (state) => {
  return state[Namespace.USER].isValidAuthorization;
};

export const getAvatar = (state) => {
  return state[Namespace.USER].avatarUrl;
};
