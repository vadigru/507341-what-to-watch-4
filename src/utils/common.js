import {RatingLevel, ALL_GENRES} from "../const.js";

export const getTextedRating = (rating) => {
  switch (true) {
    case rating >= RatingLevel.ZERO && rating <= RatingLevel.BAD:
      return `Bad`;
    case rating > RatingLevel.BAD && rating <= RatingLevel.NORMAL:
      return `Normal`;
    case rating > RatingLevel.NORMAL && rating <= RatingLevel.GOOD:
      return `Good`;
    case rating > RatingLevel.GOOD && rating < RatingLevel.AWESOME:
      return `Very good`;
    default:
      return `Awesome`;
  }
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const getMoviesByGenre = (genre, movies) => {
  return genre === ALL_GENRES
    ? movies
    : movies.filter((movie) => movie.genre === genre);
};
