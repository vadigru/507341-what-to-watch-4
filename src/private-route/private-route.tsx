import * as React from "react";
import {connect} from "react-redux";
import {Route, Redirect, RouteComponentProps} from "react-router-dom";

import {AuthorizationStatus} from "../reducer/user/user.js";
import {getAuthorizationStatus} from "../reducer/user/selectors.js";

import {AppRoute} from "../const.js";

interface Props {
  authorizationStatus: string;
  exact: boolean;
  path: string;
  render: (routeProps: RouteComponentProps<number> | null) => void;
}

const PrivateRoute: React.FunctionComponent<Props> = (props: Props) => {
  const {render, path, exact, authorizationStatus} = props;

  return (
    <Route
      path={path}
      exact={exact}
      render={(routeProps) => {
        return authorizationStatus === AuthorizationStatus.AUTH ? (
          render(routeProps)
        ) : (
          <Redirect to={AppRoute.SIGN_IN} />
        );
      }}
    />
  );
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state)
});

export {PrivateRoute};
export default connect(mapStateToProps)(PrivateRoute);
