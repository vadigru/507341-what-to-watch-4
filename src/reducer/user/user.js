import {extend} from "../../utils/common.js";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`
};

const SERVER_URL = `https://4.react.pages.academy`;

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  isValidAuthorization: true,
  avatarUrl: ``,
};

const ActionType = {
  REQUIRE_AUTHORIZATION: `REQUIRE_AUTHORIZATION`,
  IS_INVALID_AUTHORIZATION: `IS_INVALID_AUTHORIZATION`,
  SET_AVATAR_URL: `SET_AVATAR_URL`
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRE_AUTHORIZATION,
    payload: status
  }),
  requireValidAuthorization: (isValid) => ({
    type: ActionType.IS_INVALID_AUTHORIZATION,
    payload: isValid
  }),
  setAvatarUrl: (avatarUrl) => ({
    type: ActionType.SET_AVATAR_URL,
    payload: avatarUrl
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRE_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload
      });
    case ActionType.IS_INVALID_AUTHORIZATION:
      return extend(state, {
        isValidAuthorization: action.payload
      });
    case ActionType.SET_AVATAR_URL:
      return extend(state, {
        avatarUrl: `${SERVER_URL}${action.payload}`
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
    .then((response) => {
      const {avatar_url: avatar} = response.data;
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAvatarUrl(avatar));
    })
    .catch((err) => {
      throw err;
    });
  },

  login: (authData) => (dispatch, getState, api) => {
    // dispatch(ActionCreator.requireValidAuthorization(true));
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password
    })
    .then((response) => {
      const {avatar_url: avatar} = response.data;
      dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      dispatch(ActionCreator.setAvatarUrl(avatar));
    })
    .catch((err)=>{
      dispatch(ActionCreator.requireValidAuthorization(false));
      throw err;
    });
  }
};

export {ActionCreator, ActionType, AuthorizationStatus, Operation, reducer};
