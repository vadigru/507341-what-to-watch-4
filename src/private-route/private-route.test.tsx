import * as React from "react";
import * as renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {Router} from "react-router-dom";

import {PrivateRoute} from "./private-route";

import {AuthorizationStatus} from "../reducer/user/user";
import {AppRoute} from "../const";
import history from "../history";
import {noop} from "../utils/common";

it(`Should render PrivateRoute component`, () => {
  const mockStore = configureStore([]);

  const store = mockStore();

  const tree = renderer
    .create(
        <Provider store={store}>
          <Router history={history}>
            <PrivateRoute
              render={noop}
              path={AppRoute.ROOT}
              exact
              authorizationStatus={AuthorizationStatus.NO_AUTH}
            />
          </Router>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
