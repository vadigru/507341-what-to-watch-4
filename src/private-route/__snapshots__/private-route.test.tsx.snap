// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render PrivateRoute component 1`] = `null`;
